src/main/java/com/arpnetworking/metrics/mad/sources/TransformingSource.java

```diff
@@ -0,0 +1,429 @@
+                .put("findAndReplace", _findAndReplace)
+                .build();
+    }
+
+    @Override
+    public String toString() {
+        return toLogValue().toString();
+    }
+
+    private TransformingSource(final Builder builder) {
+        super(builder);
+        _source = builder._source;
+
+        final ImmutableMap.Builder<Pattern, List<String>> findReplaceBuilder =
+                ImmutableMap.builderWithExpectedSize(builder._findAndReplace.size());
+        for (final Map.Entry<String, ? extends List<String>> entry : builder._findAndReplace.entrySet()) {
+            findReplaceBuilder.put(Pattern.compile(entry.getKey()), ImmutableList.copyOf(entry.getValue()));
+        }
+        _findAndReplace = findReplaceBuilder.build();
+
+        _inject = builder._inject;
+        _remove = builder._remove;
+
+        _source.attach(new TransformingObserver(this, _findAndReplace, _inject, _remove));
+    }
+
+    private final Source _source;
+    private final ImmutableMap<Pattern, List<String>> _findAndReplace;
+    private final ImmutableMap<String, DimensionInjection> _inject;
+    private final ImmutableList<String> _remove;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TransformingSource.class);
+    private static final Splitter.MapSplitter TAG_SPLITTER = Splitter.on(';').omitEmptyStrings().trimResults().withKeyValueSeparator('=');
+
+    // NOTE: Package private for testing
+    /* package private */ static final class TransformingObserver implements Observer {
+
+        /* package private */ TransformingObserver(
+                final TransformingSource source,
+                final Map<Pattern, List<String>> findAndReplace,
+                final ImmutableMap<String, DimensionInjection> inject,
+                final ImmutableList<String> remove) {
+            _source = source;
+            _findAndReplace = findAndReplace;
+            _inject = inject;
+            _remove = remove;
+        }
+
+        @Override
+        public void notify(final Observable observable, final Object event) {
+            if (!(event instanceof Record)) {
+                LOGGER.error()
+                        .setMessage("Observed unsupported event")
+                        .addData("event", event)
+                        .log();
+                return;
+            }
+
+            // Merge the metrics in the record together
+            final Record record = (Record) event;
+            final Map<ImmutableMap<String, String>, Map<String, MergingMetric>> mergedMetrics = Maps.newHashMap();
+            for (final Map.Entry<String, ? extends Metric> metric : record.getMetrics().entrySet()) {
+                boolean found = false;
+                final String metricName = metric.getKey();
+                for (final Map.Entry<Pattern, List<String>> findAndReplace : _findAndReplace.entrySet()) {
+                    final Pattern metricPattern = findAndReplace.getKey();
```

Comment:
Can you rename this `findAndReplacePattern`?

The one thing I am not seeing here is how to select which metrics to apply which rules to. The difference between selecting a metrics and transforming it. While I realize that for metric name this can often be expressed with one regex, I suspect there is actually more to this pattern.

Specifically, I believe we'll need to provide "selection" of metrics based not only on name but also on dimension set. For example: contain key, containing key regex or containing key with value or regex of value; I don't see a point in just containing value or regex of value ... but that's probably my overly typed view of the world.

Implementing the full selection is not something you need to do now. However, I would like you to consider it in how the configuration is laid out. I would like to be able to add this without making non-backwards compatible changes to the configuration. To this end I believe we need to support a list of transformations. By default the transformation can match all, and then adding selectors can narrow those down. There is one question which is whether transformations are all applied or only some.

We can default either way, supporting the other can just be a flag away. Although I am concerned about how slippery the slope may be to a graph/tree representation (instead of a list). I would like to hear your thoughts on this one.